### PR TITLE
Unify fail_stale_trials in each Storage impl

### PR DESCRIPTION
**Pull request by knshnb**
## Motivation

Currently, `fail_stale_trials` method is implemented in each Storage class (almost same functionality).

## Description of the changes

- added `_get_stale_trial_ids` method to `BaseStorage`
- unified `fail_stale_trials` and moved to `optuna/storages/_heartbeat.py`

I'm wondering if I should remove the first underscore of `_get_stale_trial_ids`.

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@not522 Could you review this PR if you have time?

**Review by not522 — COMMENTED** [review]
Thank you for your PR!
Could you make `fail_stale_trials` of `RDBStorage` and `RedisStorage` deprecated instead of removal? Because they are public methods.
Note: `fail_stale_trials` of `BaseStorage` and `_CachedStorage` are private, therefore they can be removed.

**Comment by knshnb** [comment]
Thanks for your comment! I reverted them with deprecated messages.

**Comment by knshnb** [comment]
Thanks for your comment! I reverted them with deprecated messages.

**Comment by g-votte** [comment]
@knshnb 
Thanks for your PR!

@not522 
I'm a little worried about leaving the unused method with the standard `@deprecated` policy.

- Though both `RDBStorage` and `RedisStorage` are public, we don't assume that the user directly call `fail_stale_trials` as written [here](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.storages.RDBStorage.html#optuna.storages.RDBStorage). Breaking the backward compatibility scarcely affects the library users.
- It would be harmful for maintainability. We'd have to take care of them for years, despite that the bloated storage classes are currently problematic.

> Thank you for your PR!
> Could you make fail_stale_trials of RDBStorage and RedisStorage deprecated instead of removal? Because they are public methods.

**Comment by g-votte** [comment]
Hmm, since we have a precedent in #3323, we just may want to follow the standard `@deprecated` policy so far. We can just ignore [my previous comment](https://github.com/optuna/optuna/pull/3442#issuecomment-1089868760) at least in this PR, and discuss it later in another issue/PR.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the PR. LGTM.

**Comment by not522** [comment]
> I'm wondering if I should remove the first underscore of _get_stale_trial_ids.

That makes sense, but we need more discussion on the storage API design (#2943). I prefer keeping it private in this PR.

File: optuna/storages/_base.py [hunk 1]
```diff
@@ -724,20 +724,14 @@ def record_heartbeat(self, trial_id: int) -> None:
...
+    def _get_stale_trial_ids(self, study_id: int) -> List[int]:
+        """Get the stale trial ids of the study.
 
         Args:
             study_id:
-                ID of the related study.
-        Returns:
-            List of trial IDs of the failed trials.
+                ID of the study.
         """
```
Comment (not522) [inline optuna/storages/_base.py:733]: Could you add the `Returns` section?

**Review by not522 — CHANGES_REQUESTED** [review]

**Comment by knshnb** [comment]
@not522 Thank you for the comment! PTAL. (Please ignore the merge commits, which I made by mistake...)

**Review by not522 — APPROVED** [review]
LGTM!

**Pull request merged**